Basic NGSI-LD active measures support

jason-fox opened the pull request:
This PR offers NGSI-LD support for **active measures only** under the following configuration.


```javascript
contextBroker: {
            url: 'https://192.168.1.1:1026',
            ngsiVersion: 'ld',
            jsonLdContext: 'http://context.json-ld'
        },
```

The existing NGSI-v2 support remains as default under `ngsiVersion: 'v2' `

The NGSI-LD support matches the decision trees  for NGSI v1/v2 - where necessary a third option has been added and the necessary functionality **duplicated**.

The set of tests from NGSI v2 have been duplicated and the expectations amended where necessary.

Commands, Subscriptions and Lazy Attribute Support have not been added - disabled test stubs have been created to allow them to be tested once the code base has been updated.

jason-fox commented:
The tests cases are *copies* of the NGSI-v2 tests. The disabled tests are for commands and lazy attributes and therefore *will not work* until the remaining work is done. 

Given that the tests are all new,  I'm wondering - *for the tests only* should I:

a) Make minimal duplicate and amend changes (as is)
b) Prettify the changed test files - to reduce the scope of #753 
c) Update the tests to run using ES6 syntax - to reduce the scope of #831

jason-fox commented:
The command stubs account for the 2% drop in code coverage.

fgalan commented on CHANGES_NEXT_RELEASE:
```diff
@@ -1,4 +1,5 @@
-Add Null check within executeWithSecurity() to avoid crash (#829)
\ No newline at end of file
+Add Null check within executeWithSecurity() to avoid crash (#829)
+Add NGSIv2 metadata support to attributeAlias plugin.
```
This line seems to belong to PR #839. Thus, I understand that jason-fox:feature/ngsi-ld-measure is a branch of jason-fox:feature/add-v2-metadata.

PR #839 should be merged before, then this PR updated.

jason-fox replied:
Correct - this PR is dependent upon  and extends the metadata change.

fgalan replied:
PR #839 has been merged. This PR branch should be upgrade, then this diff will disappear.

jason-fox replied:
I've merge master which should reduce the diff - Fixed with dd81751

fgalan commented:
> Given that the tests are all new, I'm wondering - _for the tests only_ should I:
> 
> a) Make minimal duplicate and amend changes (as is)
> b) Prettify the changed test files - to reduce the scope of #753
> c) Update the tests to run using ES6 syntax - to reduce the scope of #831

Given that these .js files are entirely new, maybe it makes sense to do so.

fgalan commented:
> The command stubs account for the 2% drop in code coverage.

Not a big issue from my point of view. The drop will be eventually recovered.

jason-fox commented:
> Given that these .js files are entirely new, maybe it makes sense to do so.

Fixed  (new tests only) 40875e3 - I also corrected the copyrights for the new files to 2020

fgalan commented on lib/services/devices/deviceService.js:
```diff
@@ -143,6 +143,41 @@ function createInitialEntityHandlerNgsi2(deviceData, newDevice, callback) {
 }
 
+/**
+ * Creates the response handler for the initial entity creation request using NGSIv2.
```
```suggestion
 * Creates the response handler for the initial entity creation request using NGSI-LD.
```

jason-fox replied:
Fixed b41a18a

fgalan commented on lib/commonConfig.js:
```diff
@@ -91,7 +91,8 @@ function processEnvironmentVariables() {
              'IOTA_POLLING_DAEMON_FREQ',
-             'IOTA_MULTI_CORE'
+             'IOTA_MULTI_CORE',
+             'IOTA_JSON_LD_CONTEXT'
```
New env var should be described in documentation.

jason-fox replied:
Fixed b41a18a  - Minimal Documentation for now.

fgalan commented on lib/commonConfig.js:
```diff
@@ -91,7 +91,8 @@ function processEnvironmentVariables() {
              'IOTA_APPEND_MODE',
```
CHANGES_NEXT_RELEASE entry should be provided for this PR.

jason-fox replied:
Fixed b41a18a

fgalan commented on lib/services/devices/deviceService.js:
```diff
@@ -310,6 +348,66 @@ function executeWithSecurity(requestOptions, deviceData, callback) {
 
+
+/**
+ * Creates the initial entity representing the device in the Context Broker using NGSIv2.
```
```suggestion
 * Creates the initial entity representing the device in the Context Broker using NGSI-LD.
```

jason-fox replied:
Fixed b41a18a

fgalan commented on lib/services/devices/deviceService.js:
```diff
@@ -522,6 +621,64 @@ function updateEntityNgsi2(deviceData, updatedDevice, callback) {
+
+
+/**
+ * Updates the entity representing the device in the Context Broker using NGSIv2.
```
```suggestion
 * Updates the entity representing the device in the Context Broker using NGSIv2.
```

Three times :) Please, do an overall check for this.

jason-fox replied:
Fixed b41a18a

fgalan commented on lib/services/ngsi/ngsiService.js:
```diff
@@ -397,6 +485,83 @@ function getMetaData(typeInformation, name, metadata){
+
+
+
+function convertNGSIv2ToLD(attr){
```
JavaDoc should be provided?

jason-fox replied:
Fixed b41a18a

fgalan commented on lib/services/ngsi/ngsiService.js:
```diff
@@ -397,6 +485,83 @@ function getMetaData(typeInformation, name, metadata){
+    return obj;
+}
+
+function formatAsNGSILD(json){
```
JavaDoc should be provided?

jason-fox replied:
Fixed b41a18a

fgalan commented on lib/services/ngsi/ngsiService.js:
```diff
@@ -584,14 +741,172 @@ function sendUpdateValueNgsi2(entityName, attributes, typeInformation, token, ca
+
+
+
+
```
Big bunch of empty blank lines... intentional?

jason-fox replied:
Fixed b41a18a  (if only this tidying could be done automatically)

fgalan commented on test/.jshintrc:
```diff
@@ -30,6 +30,6 @@
     "predef":
         [
-            "describe", "beforeEach", "afterEach", "it"
+            "describe", "beforeEach", "afterEach", "it", "xdescribe", "xit"
```
I have check there are 26 xdescribe() statement and 4 xit() statement added in this PR.

I understand that the work in sucesive PRs within the NGSI-LD line will be re-enable them, isn't it?

jason-fox replied:
Yes

fgalan replied:
Ok, NTC

fgalan commented on lib/services/devices/registrationUtils.js:
```diff
@@ -329,31 +378,145 @@ function sendRegistrationsNgsi2(unregister, deviceData, callback) {
+
+
+/**
+ * Sends a Context Provider registration or unregistration request to the Context Broker using NGSIv2.
```
Another one ;)

```suggestion
 * Sends a Context Provider registration or unregistration request to the Context Broker using NGSI-LD.
```

jason-fox replied:
Fixed a027234

fgalan commented on lib/services/ngsi/ngsiService.js:
```diff
@@ -193,6 +193,88 @@ function generateNGSI2OperationHandler(operationName, entityName, typeInformatio
 
+
+/**
+ * Generate an operation handler for NGSIv2-based operations (query and update). The handler takes care of identifiying
```
```suggestion
 * Generate an operation handler for NGSI-LD-based operations (query and update). The handler takes care of identifiying
```

jason-fox replied:
Fixed a027234

fgalan commented on lib/services/ngsi/ngsiService.js:
```diff
@@ -584,14 +761,164 @@ function sendUpdateValueNgsi2(entityName, attributes, typeInformation, token, ca
+
+/**
+ * Makes an update in the Device's entity in the context broker, with the values given in the 'attributes' array. This
+ * array should comply to the NGSIv2's attribute format.
```
```suggestion
 * array should comply to the NGSI-LD's attribute format.
```

jason-fox replied:
Fixed a027234

fgalan commented on lib/services/ngsi/subscriptionService.js:
```diff
@@ -245,6 +245,58 @@ function subscribeNgsi2(device, triggers, content, callback) {
 
+
+/**
+ * Makes a subscription for the given device's entity using NGSIv2, triggered by the given attributes.
```
```suggestion
 * Makes a subscription for the given device's entity using NGSIv2, triggered by the given attributes.
```

Still too many of them... an overall review should be done.

jason-fox replied:
Fixed a027234

fgalan replied:
This one was not fixed at the end (my fault: my suggestion was not correct in my first comment). Now it should be ok:

```suggestion
 * Makes a subscription for the given device's entity using NGSI-LD, triggered by the given attributes.
```

jason-fox replied:
Fixed c70d9ed

fgalan commented on lib/services/northBound/contextServer.js:
```diff
@@ -957,6 +1177,79 @@ function handleNotificationNgsi1(req, res, next) {
     }
 }
 
+function handleNotificationNgsiLD(req, res, next) {
```
Missing Javadoc ?

jason-fox replied:
Fixed c70d9ed

jason-fox replied:
Fixed and Propagated (contextServer isn't the focus of Measures) : a5c650a

fgalan commented on lib/services/northBound/contextServer.js:
```diff
@@ -1119,6 +1412,21 @@ function queryErrorHandlingNgsi2(error, req, res, next) {
     });
 }
 
+function queryErrorHandlingNgsiLD(error, req, res, next) {
```
Missing JavaDoc?

jason-fox replied:
Fixed c70d9ed

fgalan replied:
Maybe github diff is fooling me, but it seems JavaDoc was not added here, despite the "Fixed ..." mark.

![imagen](https://user-images.githubusercontent.com/1534240/75545636-263d5700-5a27-11ea-9e33-01a74c6b7cf4.png)

Can you check, please?

jason-fox replied:
Fixed  a5c650a

fgalan commented on lib/services/northBound/contextServer.js:
```diff
@@ -1159,6 +1467,21 @@ function updateErrorHandlingNgsi2(error, req, res, next) {
     });
 }
 
+function updateErrorHandlingNgsiLD(error, req, res, next) {
```
Missing JavaDoc?

jason-fox replied:
Fixed c70d9ed

fgalan replied:
Same here

jason-fox replied:
Fixed: a5c650a
